dc/s2n-quic-dc/src/socket/recv/descriptor.rs

```diff
@@ -0,0 +1,446 @@
+                let unfilled = Self { desc: Some(desc) };
+                return Err((unfilled, err));
+            }
+        };
+
+        let segment_len = cmsg.segment_len();
+        let ecn = cmsg.ecn();
+        let desc = Filled {
+            desc,
+            offset: 0,
+            len,
+            ecn,
+        };
+        let segments = Segments {
+            descriptor: Some(desc),
+            segment_len,
+        };
+        Ok(segments)
+    }
+}
+
+impl Drop for Unfilled {
+    #[inline]
+    fn drop(&mut self) {
+        if let Some(desc) = self.desc.take() {
+            // put the descriptor back in the pool if it wasn't filled
+            desc.drop_unfilled();
+        }
+    }
+}
+
+pub struct Filled {
+    desc: Descriptor,
+    offset: u16,
+    len: u16,
+    ecn: ExplicitCongestionNotification,
+}
+
+impl fmt::Debug for Filled {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let alt = f.alternate();
+
+        let mut s = f.debug_struct("Filled");
+        s.field("id", &self.desc.id())
+            .field("remote_address", &self.remote_address().get())
+            .field("ecn", &self.ecn);
+
+        if alt {
+            s.field("payload", &self.payload());
+        } else {
+            s.field("payload_len", &self.len);
+        }
+
+        s.finish()
+    }
+}
+
+impl Filled {
+    /// Returns the ECN markings for the packet
+    #[inline]
+    pub fn ecn(&self) -> ExplicitCongestionNotification {
+        self.ecn
+    }
+
+    /// Returns the length of the payload
+    #[inline]
+    pub fn len(&self) -> u16 {
+        self.len
+    }
+
+    /// Returns `true` if the payload is empty
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.len == 0
+    }
+
+    /// Returns the remote address of the packet
+    #[inline]
+    pub fn remote_address(&self) -> &Addr {
+        // NOTE: addr_mut can't be used since the `inner` is reference counted to allow for GRO
+        self.desc.addr()
+    }
+
+    /// Returns the packet payload
+    #[inline]
+    pub fn payload(&self) -> &[u8] {
+        unsafe {
+            let ptr = self.desc.data().as_ptr().add(self.offset as _);
+            let len = self.len as usize;
+            core::slice::from_raw_parts(ptr, len)
+        }
+    }
+
+    /// Returns a mutable packet payload
+    // NOTE: this is safe since we guarantee no `Filled` references overlap
+    #[inline]
+    pub fn payload_mut(&mut self) -> &mut [u8] {
+        unsafe {
+            let ptr = self.desc.data().as_ptr().add(self.offset as _);
+            let len = self.len as usize;
+            core::slice::from_raw_parts_mut(ptr, len)
+        }
+    }
+
+    /// Splits the packet into two at the given index.
+    ///
+    /// Afterwards `self` contains elements `[at, len)`, and the returned
+    /// [`Filled`] contains elements `[0, at)`.
+    ///
+    /// This is an `O(1)` operation that just increases the reference count and
+    /// sets a few indices.
```

Comment:
Stale comment? I don't see reference count updates.

Reply:
They're done in the `clone_filled` function